### PR TITLE
bpo-42269: Add slots parameter to dataclass decorator

### PR DESCRIPTION
**Pull request by uriyyo**
https://bugs.python.org/issue42269

### CODE REVIEW

**Comment by uriyyo** [comment]
Hi @ericvsmith, could you please review this PR?)

File: Doc/library/dataclasses.rst [hunk 1]
```diff
@@ -173,6 +173,10 @@ Module-level decorators, classes, and functions
      glossary entry for details.  Also see the ``dataclasses.KW_ONLY``
      section.
 
+   - ``slots``: If true (the default is ``False``), :attr:`__slots__` attribute
+     will be generated. If :attr:`__slots__` is defined in the class, then
+     :exc:`TypeError` is raised.
```
Comment (ericvsmith) [inline Doc/library/dataclasses.rst]: Could you mention here that if slots is specified then a new class is created and returned? I guess we should also add somewhere that this is the only case where a new class is created, in all other cases the original class is returned by the decorator.

File: Lib/test/test_dataclasses.py [hunk 2]
```diff
@@ -2781,6 +2781,49 @@ class Derived(Base):
...
+            y: int
+
+        self.assertEqual(Delivered.__slots__, ('x', 'y'))
+
+        @dataclass
+        class AnotherDelivered(Base):
+            z: int
+
+        self.assertTrue('__slots__' not in AnotherDelivered.__dict__)
+
```
Comment (ericvsmith) [inline Lib/test/test_dataclasses.py:2825]: And could you add a test here that a new class is returned? Something like (untested):
```
def test_returns_new_class(self):
    class A:
        x: int

    B = dataclass(A, slots=True)
    self.assertNotIs(A, B)
    # Maybe also test that A.__slots__ isn't set, and B.__slots__ is.
```

**Review by ericvsmith — CHANGES_REQUESTED** [review]
Thanks for fixing the conflicts, @uriyyo!

File: Lib/test/test_dataclasses.py (hunk 2 above)
Comment (uriyyo) [reply Lib/test/test_dataclasses.py:2825]: Done, thanks for pointing this out

**Comment by uriyyo** [comment]
I have made the requested changes; please review again

**Comment by uriyyo** [comment]
@ericvsmith Could you please verify updated docs?

Should we mention this changes at python 3.10 whats new section?

File: Doc/library/dataclasses.rst [hunk 3]
```diff
@@ -173,6 +173,11 @@ Module-level decorators, classes, and functions
      glossary entry for details.  Also see the ``dataclasses.KW_ONLY``
      section.
 
+   - ``slots``: If true (the default is ``False``), :attr:`__slots__` attribute
+     will be generated and new class will be returned instead of original one.
```
Comment (ericvsmith) [inline Doc/library/dataclasses.rst]: Please change to "instead of the original one" (add "the").

File: Doc/library/dataclasses.rst [hunk 4]
```diff
@@ -173,6 +173,11 @@ Module-level decorators, classes, and functions
...
+     If :attr:`__slots__` is defined in the class, then :exc:`TypeError`
```
Comment (ericvsmith) [inline Doc/library/dataclasses.rst]: Please change to "is already defined in the class" (add "already").

**Review by ericvsmith — CHANGES_REQUESTED** [review]
I have a few minor grammar nits, otherwise this looks good.

A whats new section for dataclasses is probably a good idea. I also need to add kw_only and match_args, too. I'll add kw_only and match_args after this PR lands.

I hope to get this all committed this weekend so it makes beta1.

File: Doc/library/dataclasses.rst (hunk 3 above)
Comment (uriyyo) [reply Doc/library/dataclasses.rst]: Done

File: Doc/library/dataclasses.rst (hunk 4 above)
Comment (uriyyo) [reply Doc/library/dataclasses.rst]: Done

**Comment by uriyyo** [comment]
I have made the requested changes; please review again

**Comment by uriyyo** [comment]
Regarding the `Whats new`. Will you mention the slots feature or should I do it?

**Comment by ericvsmith** [comment]
You should add slots, I’ll add the others in a different PR.

**Review by ericvsmith — APPROVED** [review]

**Comment by uriyyo** [comment]
@ericvsmith Could you please review `whats new`?

**Comment by ericvsmith** [comment]
Thanks, @uriyyo. This all looks good. I'll take one last pass at reviewing it this weekend and commit it.

**Pull request merged**